File: gaia-sdk-javascript/src/http/HMACTokenBuilder.ts
// @ts-ignore
import * as CryptoJS from 'crypto-js';
import {ClientOptions} from "..";
import {HMACCredentials} from "../api/GaiaCredentials";
import {HttpClient} from "./HttpClient";

export class HMACTokenBuilder {

    private payload: any;
    private nonce: string;
    private clientOptions: ClientOptions;
    private timestamp: number;

    constructor() {
        this.payload = ""
        this.nonce = ""
        this.timestamp = 0
        this.clientOptions = new ClientOptions(new HMACCredentials("",""),"application/json")
    }

    public withPayload(payload: any): HMACTokenBuilder{
        this.payload=payload
        return this
    }
    public withNonce(nonce: string): HMACTokenBuilder{
        this.nonce=nonce
        return this
    }
    public withClientOptions(clientOptions: ClientOptions): HMACTokenBuilder{
        this.clientOptions=clientOptions
        return this
    }

    public withTimestamp(timestamp: number): HMACTokenBuilder{
        this.timestamp=timestamp
        return this
    }

    /**
     * Authorization: "HMAC-SHA512 " + API_KEY + "_" +
     * base64(hmac-sha512( content, content_type, sensor_type, timestamp, nonce )) + "_" + timestamp + "_" + nonce
     */
    public async build(): Promise<string> {
        const credentials = this.clientOptions.credentials as HMACCredentials
        const sep = "_"
        const HTTP_SENSOR_TYPE = "http"
        const base64EncodedPayload = await this.toBase64(this.payload);
        let prepareToHash = [base64EncodedPayload,this.clientOptions.contentType,HTTP_SENSOR_TYPE,this.timestamp,this.nonce].join(sep)
        const hmac = CryptoJS.HmacSHA512(Buffer.from(prepareToHash).toString(),credentials.apiSecret).toString()
        const signature = await this.toBase64(hmac);

        return "HMAC-SHA512 " + [credentials.apiKey,signature,this.timestamp,this.nonce].join(sep)
    }

    private async toBase64(data: any): Promise<string> {
        if (this.isBuffer(data)) {
            // Node
            return data.toString('base64');
        } else if (this.isBrowserBlob(data)) {
            // Browser
           const arrBuff = await data.arrayBuffer();
           return new Buffer(arrBuff).toString('base64')
        } else {
            let stringData = HttpClient.asString(data)
            return typeof btoa !== 'undefined' && this.b64EncodeUnicode(stringData) || Buffer.from(this.encodeWithFixedSpecialChars(stringData), 'binary').toString('base64');
        }
    }

    // see https://stackoverflow.com/questions/30106476/using-javascripts-atob-to-decode-base64-doesnt-properly-decode-utf-8-strings
    private b64EncodeUnicode(str: any) {
        // first we use encodeURIComponent to get percent-encoded UTF-8,
        // then we convert the percent encodings into raw bytes which
        // can be fed into btoa.
        return btoa(this.encodeWithFixedSpecialChars(str));
    }

    private encodeWithFixedSpecialChars(str: any) {
        return encodeURIComponent(str).replace(/%([0-9A-F]{2})/g,
            function toSolidBytes(match, p1) {
                return String.fromCharCode(Number('0x' + p1));
            });
    }

    private isBuffer(b: Buffer): b is Buffer {
        return !!(b as Buffer).writeFloatLE;
    }

    private isBrowserBlob(blob: any): boolean {
        return typeof(blob) === 'object' && blob.toString() === '[object Blob]';
    }
}

File: gaia-sdk-python/gaia_sdk/http/HMACTokenBuilder.py
from gaia_sdk.api.crypto import HMAC
import base64
import json


class HMACTokenBuilder(object):

    def with_client_options(self, client_options):
        self.client_options = client_options
        return self

    def with_nonce(self, nonce):
        self.nonce = nonce
        return self

    def with_payload(self, payload):
        self.payload = payload
        return self

    def with_timestamp(self, timestamp):
        self.timestamp = timestamp
        return self

    def build(self):
        """
        Authorization: "HMAC-SHA512 " + API_KEY + "_" +
        base64(hmac-sha512( content, content_type, sensor_type, timestamp, nonce )) + "_" + timestamp + "_" + nonce
        """
        HTTP_SENSOR_TYPE = "http"
        sep = "_"

        if isinstance(self.payload, bytes):
            data_to_encode = self.payload
        else:
            data_to_encode = json.dumps(self.payload).encode("utf-8")

        base64_string = base64.b64encode(data_to_encode).decode()

        arrayToHash = [base64_string, self.client_options.content_type,
                       HTTP_SENSOR_TYPE, self.timestamp, self.nonce]
        prepareToHash = '_'.join([str(x) for x in arrayToHash])
        hmac = HMAC(self.client_options.credentials.apiSecret)

        signature = hmac.hash512(prepareToHash)
        token = "HMAC-SHA512 " + self.client_options.credentials.apiKey + sep + signature + sep + str(
            self.timestamp) + sep + self.nonce
        return token
